llm/ext_server.go

```diff
@@ -0,0 +1,394 @@
+	return C.llama_server_completion_cancel(task_id)
+}
+func (llm *llamaExtServer) llama_server_tokenize(json_req *C.char, resp *C.ext_server_resp) C.ext_server_err {
+	return C.llama_server_tokenize(json_req, resp)
+}
+func (llm *llamaExtServer) llama_server_detokenize(json_req *C.char, resp *C.ext_server_resp) C.ext_server_err {
+	return C.llama_server_detokenize(json_req, resp)
+}
+func (llm *llamaExtServer) llama_server_embedding(json_req *C.char, resp *C.ext_server_resp) C.ext_server_err {
+	return C.llama_server_embedding(json_req, resp)
+}
+
+func newLlamaExtServer(model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+	server := &llamaExtServer{opts}
+	log.Printf("Loading CUDA llm server")
+	return newExtServer(server, model, adapters, projectors, numLayers, opts)
+}
+
+func newExtServer(server extServer, model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+	if !mutex.TryLock() {
+		log.Printf("concurrent llm servers not yet supported, waiting for prior server to complete")
+		mutex.Lock()
+	}
+	fileInfo, err := os.Stat(model)
+	if err != nil {
+		return nil, err
+	}
+	var sparams C.ext_server_params
+	sparams.model = C.CString(model)
+	defer C.free(unsafe.Pointer(sparams.model))
+
+	numGPU := gpu.NumGPU(numLayers, fileInfo.Size(), opts)
+
+	sparams.embedding = true
+	sparams.n_ctx = C.uint(opts.NumCtx)
+	sparams.n_batch = C.uint(opts.NumBatch)
+	sparams.n_gpu_layers = C.int(numGPU)
+	sparams.main_gpu = C.int(opts.MainGPU)
+	sparams.n_parallel = 2 // TODO - wire up concurrency
+
+	// Always use the value encoded in the model
+	sparams.rope_freq_base = 0.0
+	sparams.rope_freq_scale = 0.0
+
+	sparams.lora_adapters = nil
+	for i := 0; i < len(adapters); i++ {
+		la := (*C.ext_server_lora_adapter)(C.malloc(C.sizeof_struct_ext_server_lora_adapter))
+		defer C.free(unsafe.Pointer(la))
+		la.adapter = C.CString(adapters[i])
+		defer C.free(unsafe.Pointer(la.adapter))
+		la.scale = C.float(1.0) // TODO expose scale/weights up through ollama UX
+		la.next = nil
+		if i == 0 {
+			sparams.lora_adapters = la
+		} else {
+			tmp := sparams.lora_adapters
+			for ; tmp.next != nil; tmp = tmp.next {
+			}
+			tmp.next = la
+		}
+	}
+
+	// TODO - implement ME
+	// if len(projectors) > 0 {
+	// 	// TODO: applying multiple projectors is not supported by the llama.cpp server yet
+	// 	params = append(params, "--mmproj", projectors[0])
+	// }
+
+	if opts.NumThread > 0 {
+		sparams.n_threads = C.uint(opts.NumThread)
+	} else {
+		sparams.n_threads = C.uint(runtime.NumCPU())
+	}
+
+	sparams.memory_f16 = false
+	if opts.F16KV {
+		sparams.memory_f16 = true
+	}
+	sparams.use_mlock = false
+	if opts.UseMLock {
+		sparams.use_mlock = true
+	}
+	sparams.use_mmap = true
+	if !opts.UseMMap {
+		sparams.use_mmap = false
+	}
+	sparams.numa = false
+	if opts.UseNUMA {
+		sparams.numa = true
+	}
```

Comment:
Can we just set these directly rather than initializing to false and checking the conditional?
```suggestion
	sparams.memory_f16 = C.bool(opts.F16KV)
	sparams.use_mlock = C.bool(opts.UserMLock)
	sparams.use_mmap = C.bool(opts.UserMMap)
	sparams.numa = C.bool(opts.UseNUMA)
```